docs/architecture/adr-074-timeout-params.md

```diff
@@ -0,0 +1,228 @@
+# ADR 74: Migrate Timeout Parameters to Consensus Parameters
+
+## Changelog
+
+- 03-Jan-2022: Initial draft (@williambanfield)
+
+## Status
+
+Proposed
+
+## Context
+
+### Background
+
+Tendermint's consensus timeout parameters are currently configured locally by each validator
+in the validator's [config.toml][config-toml].
+This means that the validators on a Tendermint network may have different timeouts
+from each other. There is no reason for validators on the same network to configure
+different timeout values. Proper functioning of the Tendermint consensus algorithm
+relies on these parameters being uniform across validators.
+
+The configurable values are as follows:
+
+* `TimeoutPropose` 
+	* How long the consensus algorithm waits for a proposal block before issuing a prevote.
+	* If no prevote arrives by `TimeoutPropose`, then the consensus algorithm will issue a nil prevote.
+* `TimeoutProposeDelta` 
+	* How much the `TimeoutPropose` grows each round.
+* `TimeoutPrevote` 
+	* How long the consensus algorithm waits after receiving +2/3 prevotes before 
+	issuing a precommit.
+* `TimeoutPrevoteDelta` 
+	* How much the `TimeoutPrevote` increases with each round.
+* `TimeoutPrecommit` 
+	* How long the consensus algorithm waits after receiving +2/3 precommits with 
+	no quorum before entering the next round.
+* `TimeoutPrecommitDelta` 
+	* How much the `TimeoutPrecommit` increases with each round.
+* `TimeoutCommit` 
+	* How long the consensus algorithm waits after committing a block but before starting the new height.
+	* This gives a validator a chance to receive slow precommits.
+* `SkipTimeoutCommit`
+	* Make progress as soon as the node has 100% of the precommits.
+
+
+### Overview of Proposed Change
+
+This ADR proposes consolidating the timeout parameters and migrating them from the
+node-local `config.toml` file into the network-global consensus parameters.
+The consensus parameters are the same across all validators on the network.
+
+A sane default will be provided by Tendermint for each of these parameters and
+networks will be able to update the parameters as they see fit. Local updates
```

Comment:
Have you thought about what the logistics are around an upgrade? Is there any possibility that the application could set the defaults or must they take our defaults and only change them after a governance proposal is passed?

Reply:
> Is there any possibility that the application could set the defaults or must they take our defaults and only change them after a governance proposal is passed?

This is an important question, I'll do a bit of research to see what options may exist to allow application developers to pick values ahead of the change.

Reply:
Applications could bypass the governance mechanism by returning values at a specific height. The application would have to maintain the logic to add something at a certain height forever to allow validators that join the chain after the change has taken effect to receive the param change.

Reply:
Added a migration section :+1:

Reply:
I guess this would also apply to PBTS so having an appropriate mechanism for application developers to set the initial values seems like a good idea. Have you thought perhaps about using `InitChain` upon upgrade to fetch the new value?

Reply:
I think that re-purposing that call is pretty risky. We currently document that call as only being made on chain-start and we don't provide a signal for applications to actually determine the height at which the call is being made so it's not obvious how developers could distinguish between 'chain start' and 'upgrade' 

This also would not be present for nodes that are syncing to the network. If the upgrade happened at height 100 and a node joined after the upgrade, there is not obvious way for that application to learn that a consensus param changed at height 100. This is, imo, why keeping it in EndBlock is pretty reasonable.

This does leave at least one height that must use the defaults, i.e. the first height after the upgrade where the new params exist but are not yet set. Given that the default params do not appear to be changed by popular chains makes this seem very low risk, but I definitely see it as pretty inelegant. This is a good candidate for discussion in our next dev session. I'll add it to the agenda there.

Reply:
> it's not obvious how developers could distinguish between 'chain start' and 'upgrade'

I was thinking of the case where it's one and the same binary in which case the developers don't really need to know what height it is at.

I think you're probably right about it not being worth it, I just think that this might be a common occurrence and perhaps we develop something specifically for that use case.

Reply:
Yeah, this is a pretty complex thing to do. Captured thoughts here: https://github.com/tendermint/tendermint/pull/7524